File: ch1/BasicCoinChangingStart.py
#
# From: "Chains that bind us" by Phillip G. Bradford
#  https://github.com/wonder-phil/ChainsThatBindUs
#     
import numpy as np

class GreedyChange:

    def __init__(self, denominations, givenTotal):
        self.denominations = np.flip(np.sort(denominations)) # smallest to largest
        self.coinAmounts = [0 for i in range(len(self.denominations))]
        self.givenTotal = givenTotal

    def computeChange(self):
        total = self.givenTotal
        indexLargestRemaining = 0
        while total > 0:
            self.coinAmounts[indexLargestRemaining] = total // self.denominations[indexLargestRemaining]
            total = self.givenTotal % self.denominations[indexLargestRemaining]
            indexLargestRemaining = indexLargestRemaining + 1


    def __str__(self):
        s = 'Given total: '+ str(self.givenTotal) + '\n'
        s = s + 'Denominations: ' + str(self.denominations) + '\n'
        s = s + 'Numbers of coins using greedy' + '\n'
        s = s + '-----------------------------\n'
        for (amount, coin) in zip(self.coinAmounts, self.denominations):
            s = s + str(amount) + ' of ' + str(coin) + '\n'
            
        return s


    

File: cpp/Ch1/BasicCoinChangingStart.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class GreedyChange {
private:
    vector<int> denominations;
    vector<int> coinAmounts;
    int givenTotal;
public:
    GreedyChange(vector<int> denominations, int givenTotal) {
        sort(denominations.begin(), denominations.end(), greater<int>());
        this->denominations = denominations;
        this->coinAmounts = vector<int>(denominations.size(), 0);
        this->givenTotal = givenTotal;
    }
    void compute_change() {
        int total = givenTotal;
        int indexLargestRemaining = 0;
        while (total > 0) {
            coinAmounts[indexLargestRemaining] = total / denominations[indexLargestRemaining];
            total = givenTotal % denominations[indexLargestRemaining];
            indexLargestRemaining++;
        }
    }
    string toString() {
        string s = "Given total: " + to_string(givenTotal) + "\n";
        s += "Denominations: ";
        for (int i = 0; i < denominations.size(); i++) {
            s += to_string(denominations[i]) + " ";
        }
        s += "\n";
        s += "Numbers of coins using greedy\n";
        s += "-----------------------------\n";
        for (int i = 0; i < coinAmounts.size(); i++) {
            s += to_string(coinAmounts[i]) + " of " + to_string(denominations[i]) + "\n";
        }
        s += "\n";
        return s;
    }
};

int main() {
    vector<int> denominations = {1, 5, 10, 25};
    int givenTotal = 67;
    GreedyChange gc(denominations, givenTotal);
    gc.compute_change();
    cout << gc.toString();
    return 0;
}

